### PR TITLE
diffusion using io-classes

### PR DESCRIPTION
**Pull request by coot**
This PR allows to run diffusion in io-sim.  It also refactors the p2p
/ non-p2p: using records instead of functions with many arguments and enables
to have simpler `cardano-node` integration.

- diffusion-policies: removed dead code
- ouroboros-network: various stylystic changes
- snocket: support getLocalName for named pipes (Windows)
- netenv: removed nsBoundAddresses
- netenv: more descriptive error messages
- handshake: remove versioned application from HandshakeArguments
- peer-selection: polymorphic address type
- diffusion: generalise to an abstract peer address
- diffusion: run diffusion in an abstract monad
- diffusion: simplify data diffusion
- node-to-client: export LocalSocket constructor

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (bolt12) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:70]: Is this comment necessary?

**Review by bolt12 — APPROVED** [review]
LGTM! This should also have a corresponding PR in `cardano-node` since you changed the Diffusion Layer

**Comment by bolt12** [comment]
Also it seems CI is not passing

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: removed

**Comment by coot** [comment]
> LGTM! This should also have a corresponding PR in cardano-node since you changed the Diffusion Layer

I almost have it, just need to test if the node can run in both p2p and non-p2p mode.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ImmutableDB/Impl/Index/Primary.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ImmutableDB/Impl/Index/Primary.hs:644]: Is this for `hasktags` or some such? If it's just an aesthetics change, please remove it.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:716]: Changing this data type changes the way that `cardano-node` invokes the Consensus layer. So: be advised they might need guidance about how to use these two new arguments etc.

**Review by nfrisby — DISMISSED** [review]
I reviewed the Consensus changes. Thanks for integrating it 👍 

I Requested Changes only because I think that's an unnecessary change and it's the only change in that file; so the git history would be cleaner if we avoid that diff.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ImmutableDB/Impl/Index/Primary.hs
Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ImmutableDB/Impl/Index/Primary.hs:644]: `ghcid` tripped over it

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:716]: I have a corresponding pr here: https://github.com/input-output-hk/cardano-node/pull/2950

Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:716]: btw, both are to `p2p-master` branch

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ImmutableDB/Impl/Index/Primary.hs
Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ImmutableDB/Impl/Index/Primary.hs:644]: I can put this in a separate commit; I didn't intended to commit this, but it would be quite useful for the networking team.  Working with network integration is hard because building `lib:ouroboros-consensus` after a simple change in `ouroboros-network` requires re-building > 200 modules which are independent of `ouroboros-netowrk`. `ghcid` allows to mitigate this.  We have an issue #2614 to split the consensus into smaller logical units which I would like to resurrect once we merge `p2p-master` to `master`.

Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ImmutableDB/Impl/Index/Primary.hs:644]: Other than that, I agree that `(!)` notation is ugly.

Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ImmutableDB/Impl/Index/Primary.hs:644]: The error I am getting is:
```
src/Ouroboros/Consensus/Storage/ImmutableDB/Impl/Index/Primary.hs:643:1-3: error:
    The type signature for ‘!’ lacks an accompanying binding
      (The type signature must be given where ‘!’ is declared)
    |
643 | (!) :: (HasCallStack, V.Unbox a) => Vector a -> Int -> a
    | ^^^
src/Ouroboros/Consensus/Storage/ImmutableDB/Impl/Index/Primary.hs:650:12-14: error:
    The INLINE pragma for ‘!’ lacks an accompanying binding
      (The INLINE pragma must be given where ‘!’ is declared)
    |
650 | {-# INLINE (!) #-}
    |            ^^^
```

I think that's caused by `:set -XBangPatterns` in my `~/.ghci`; I'll remove it, and clean the commit.

Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ImmutableDB/Impl/Index/Primary.hs]: Huh! I've never come across that. Seems like a GHC bug to me, on first thought -- but many GHC things require more than the first thought...

Thanks for adjusting your setup.

**Review by nfrisby — COMMENTED** [review]
Thanks for addressing my concern. Dismissing my Request Changes review so that Armando's Approval is unmasked.

**Pull request merged**
